### ISSUE TITLE
Why status 422 instead of 400 when invalid (header, query, etc.) parameters in the request?

### ISSUE THREAD

**Issue by yiannis-kt** [tiangolo/fastapi#643]
**Description**

In the case of an invalid parameter in the request (e.g. header, query, etc.) fastapi returns a 422 status code with the following example body:

```
{
  "detail": [
    {
      "loc": [
        "header",
        "foo-bar"
      ],
      "msg": "field required",
      "type": "value_error.missing"
    }
  ]
}
```

Firstly, I was wondering what are the reasons for such choice instead of the most common 400 status code. It seems to me a bit strange, especially in the case where a required header is missing, as I would think this classifies as a syntactic error (from description of [422](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/422) versus [400](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400) codes).

As such, is there any easy way to change default return http code on validation errors, while ideally keeping the same return body whose structure is pretty neat, instead of defining a custom schema as kind of described in issue #429 ?

**Comment by yusra-haider** [comment]
from what I understand, following this answer https://github.com/tiangolo/fastapi/issues/553#issuecomment-533644801 retains the response structure while changing the status code, without defining an elaborate custom schema. It doesn't update the status codes in the automatically generated docs though...

**Comment by tiangolo** [comment]
The difference between 422 and 400 becomes a bit subjective.

The rationale is that the error is not about using the protocols and languages incorrectly (HTTP, JSON) nor about doing something otherwise not allowed or incorrect (like performing an operation without enough privileges), but about sending invalid contents, even though using the correct content format (JSON) through the correct channel (valid HTTP).

But I acknowledge that the election between 422 vs others is interpretable and subjective.

The second reason is that Flask-apispec with Marshmallow used 422 too, I thought that was a good idea.

And the other advantage is that you can then separate a whole class of errors, validation errors, that are created automatically by your FastAPI app. And handle them all in a consistent way in your client, without having to do some extra inspections of the contents to decide if it's a validation error or not.

And then you can raise 400 errors in your own code, and know that those 400 errors will be something specific that you raise, and handle them in a different, custom way in your client.

You can override the validation errors including the status code, here are the docs: https://fastapi.tiangolo.com/tutorial/handling-errors/#override-request-validation-exceptions

**Comment by yusra-haider** [comment]
@tiangolo The only issue with this approach in the docs is that auto-generated documentation doesn't reflect the updated status code. The second issue, perhaps a more pedantic one, is that I also have to define the response structure again, when all I want to do is change the status code.

There's a very straightforward way to change the status code in webargs: https://webargs.readthedocs.io/en/latest/advanced.html#returning-http-400-responses

Is there any way we could have a way to define a custom status code for validation exceptions, and have that reflected in the generated docs, without defining and raising custom exceptions?

**Comment by yiannis-kt** [comment]
Thanks @tiangolo and @yusra-haider for your replies. Sound good.

**Comment by tiangolo** [comment]
@yusra-haider yep. Check the docs there, in the same link as above. You can add a custom exception handler without creating a custom exception: https://fastapi.tiangolo.com/tutorial/handling-errors/#use-the-requestvalidationerror-body

Thanks @yiannis-kt for reporting back and closing the issue :+1:

**Comment by renzodgc** [comment]
Hey @tiangolo , sorry for reviving this. But I'm looking for a way of changing the status code in the generated docs (I changed the `validation_exception_handler` so that it returns 400 and the body of the request)

According to your last message I understand there is a way of achieving this, but I'm not being able to understand how by reading the section you linked. Is it indeed possible?

**Comment by PhilippeGalvan** [comment]
Hey @tiangolo, I am also trying to reflect the updated status_code into the auto-doc without success, fastapi is really nice and I'm trying to convince a client to use it and auto-doc is a key argument !

```python
@app.exception_handler(RequestValidationError)
async def validation_exception_handler(
    request: Request,
    exc: RequestValidationError
):
    return JSONResponse(
        status_code=status.HTTP_400_BAD_REQUEST,
        content=jsonable_encoder({"detail": exc.errors(), "body": exc.body}),
    )
```

Btw, thanks a lot for this framework and your dedication :) !

**Comment by Kludex** [comment]
A discussion about a solution is happening here #1376

**Comment by antonagestam** [comment]
> even though using the correct content format (JSON) through the correct channel (valid HTTP)

This is the problem, the format here is not simply JSON, but a rich OpenAPI schema. I see no reason why violating that schema, which is clearly the contract for interacting with the API, is not a client error. Of course one can argue this is subjective, but interpreting sending a string instead of a number as "the syntax of the request entity is correct" seems like a far fetch.

The reason that this is controversial to me is that 422 is a very nice (relatively) new response code to be able to use and distinguish from 400 (which always indicates there's a bug in the client) as meaning that the entity is in a state where it cannot currently be processed.

**Comment by Jaza** [comment]
I don't mind the fact that FastAPI returns 422 when the request body contains syntactically valid JSON but fails pydantic validation. However, I think that it really should return 400 when the request body contains syntactically invalid JSON. This is what I came up with to achieve that behaviour:

```python
from typing import Any, Sequence

from fastapi import FastAPI, Request, status
from fastapi.encoders import jsonable_encoder
from fastapi.exceptions import RequestValidationError
from fastapi.responses import JSONResponse


def is_error_dict_bad_request(errors: Sequence[Any]) -> bool:
    """Check if the given error indicates a malformed request."""
    if not len(errors) == 1:
        return False

    error_item = errors[0]

    if not isinstance(error_item, dict):
        return False

    if not isinstance(error_item.get("loc"), list):
        return False

    loc = error_item["loc"]

    if not 1 <= len(loc) <= 2:
        return False

    loc_item1 = loc[0]

    if loc_item1 != "body":
        return False

    loc_item2 = loc[1] if len(loc) > 1 else None

    if loc_item2:
        return False

    if not isinstance(error_item.get("msg"), str):
        return False

    msg = error_item["msg"]

    return (
        msg == "field required"
        or msg == "value is not a valid dict"
        or msg.startswith("Expecting value:")
    )


app = FastAPI()


@app.exception_handler(RequestValidationError)
async def request_validation_error_handler(
    request: Request, exc: RequestValidationError
) -> JSONResponse:
    errors = jsonable_encoder(exc.errors())
    status_code = (
        status.HTTP_400_BAD_REQUEST
        if is_error_dict_bad_request(errors)
        else status.HTTP_422_UNPROCESSABLE_ENTITY
    )

    return JSONResponse(
        status_code=status_code,
        content={"detail": errors},
    )
```

Then, it returns the appropriate status code for each use case.

Wrong content type in request header:

```
curl -X POST \
  -H "Content-Type: text/plain" \
  -d "{\"foo\": \"woo\"}" -v \
  http://localhost:8000/foo/
< HTTP/1.1 400 Bad Request
{
  "detail": [
    {
      "loc": [
        "body"
      ],
      "msg": "value is not a valid dict",
      "type": "type_error.dict"
    }
  ]
}
```

Missing request body:

```
curl -X POST \
  -H "Content-Type: application/json" -v \
  http://localhost:8000/foo/
< HTTP/1.1 400 Bad Request
{
  "detail": [
    {
      "loc": [
        "body"
      ],
      "msg": "field required",
      "type": "value_error.missing"
    }
  ]
}
```

Syntactically invalid JSON in request body:

```
curl -X POST \
  -H "Content-Type: application/json" \
  -d "foo" -v \
  http://localhost:8000/foo/
< HTTP/1.1 400 Bad Request
{
  "detail": [
    {
      "loc": [
        "body",
        0
      ],
      "msg": "Expecting value: line 1 column 1 (char 0)",
      "type": "value_error.jsondecode",
      "ctx": {
        "msg": "Expecting value",
        "doc": "foo",
        "pos": 0,
        "lineno": 1,
        "colno": 1
      }
    }
  ]
}
```

Syntactically valid JSON that fails pydantic validation in request body:

```
curl -X POST \
  -H "Content-Type: application/json" \
  -d "{\"foo\": \"woo\"}" -v \
  http://localhost:8000/foo/
< HTTP/1.1 422 Unprocessable Entity
{
  "detail": [
    {
      "loc": [
        "body",
        "moo"
      ],
      "msg": "field required",
      "type": "value_error.missing"
    }
  ]
}
```
